microprofile/server/src/main/java/io/helidon/microprofile/server/ServerCdiExtension.java

```diff
@@ -216,6 +228,17 @@ private void registerClasspathStaticContent(Config config) {
     }
 
     private void stopServer(@Observes @Priority(PLATFORM_BEFORE) @BeforeDestroyed(ApplicationScoped.class) Object event) {
+        boolean isRunning = IN_PROGRESS_OR_RUNNING.get();
+        try {
+            doStop(event);
+        } finally {
+            if (isRunning) {
+                IN_PROGRESS_OR_RUNNING.set(false);
+            }
```

Comment:
This logic sets `IN_PROGRESS_OR_RUNNING` to false only if it was true at the start of the method. Could there be a race condition here in which an (incorrect) second attempt to start the server concurrently occurred after `isRunning` is assigned `false` (and that parallel start-up would set `IN_PROGRESS_OR_RUNNING` to `true`), the `doStop` actually does the shutdown, but because `isRunning` is not true then `IN_PROGRESS_OR_RUNNING` is not set to `false`? So the server would be stopped but `IN_PROGRESS_OR_RUNNING` incorrectly says `true`. 

Maybe this cannot happen in practice because of other reasons?

Reply:
There was a problem with this. Fixing.

Reply:
Fixing this uncovered other issues with correct shutdown. Fixed them as well.